pkg/controller/controller.go

```diff
@@ -352,6 +360,23 @@ func verifyNoCrossNamespaceRefs(canary *flaggerv1.Canary) error {
 	return nil
 }
 
+func verifyKnativeCanary(canary *flaggerv1.Canary) error {
+	if canary.Spec.TargetRef.Kind == "Service" && canary.Spec.TargetRef.APIVersion == "serving.knative.dev/v1" {
+		if canary.Spec.Provider != flaggerv1.KnativeProvider {
+			return fmt.Errorf("can't use %s provider with Knative Service as target", canary.Spec.Provider)
+		}
+	}
+
+	if canary.Spec.Provider == flaggerv1.KnativeProvider {
+		if canary.Spec.TargetRef.Kind != "Service" || canary.Spec.TargetRef.APIVersion != "serving.knative.dev/v1" {
```

Comment:
We have this check in lots of places `canary.Spec.TargetRef.APIVersion != "serving.knative.dev/v1"`. I would create a function e.g. `IsKnativeAPI(apiVersion string) bool` and call it everywhere where the string `serving.knative.dev/v1` is currently used.